<a href="https://colab.research.google.com/github/lukwac123/machine_learning_bootcamp/blob/main/supervised/02_regression/01_linear_regression_normal_quation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np

X1 = np.array([1, 2, 3, 4, 5, 6])
Y = np.array([3000, 3250, 3500, 3750, 4000, 4250])
m = len(X1)

print(f'Lata pracy: {X1}')
print(f'Wynagrodzenie: {Y}')
print(f'Liczba próbek: {m}')

Lata pracy: [1 2 3 4 5 6]
Wynagrodzenie: [3000 3250 3500 3750 4000 4250]
Liczba próbek: 6


In [2]:
X1 = X1.reshape(m, 1)
print(X1)
print(X1.shape)

[[1]
 [2]
 [3]
 [4]
 [5]
 [6]]
(6, 1)


In [3]:
bias = np.ones((m, 1))
print(bias)
print(bias.shape)

[[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]
(6, 1)


In [4]:
X = np.append(bias, X1, axis=1)
print(X)
print(X.shape)

[[1. 1.]
 [1. 2.]
 [1. 3.]
 [1. 4.]
 [1. 5.]
 [1. 6.]]
(6, 2)


# Równanie normalne - równanie pozwalające obliczyć minimum funkcji straty (o ile istnieje)
## W = (X<sup>T</sup>X)<sup>-1</sup>(X<sup>T</sup>Y)



In [5]:
np.dot(X.T, X)

array([[ 6., 21.],
       [21., 91.]])

In [6]:
L = np.linalg.inv(np.dot(X.T, X))
L

array([[ 0.86666667, -0.2       ],
       [-0.2       ,  0.05714286]])

In [7]:
P = np.dot(X.T, Y)
P

array([21750., 80500.])

In [8]:
np.dot(L, P)

array([2750.,  250.])

# Końcowa postać modelu
## Y = 2750+250X<sub>1</sub>

# Regresja liniowa przy pomocy scikit-learn

In [9]:
from sklearn.linear_model import LinearRegression

regression = LinearRegression()
regression.fit(X1, Y)

print(regression.intercept_)
print(regression.coef_[0])

2750.0
250.0
